pkg/snapshot/controllers/snapshotrestore.go

```diff
@@ -45,34 +53,165 @@ func (c *SnapshotRestoreController) Init() error {
 
 // Handle updates for SnapshotRestore objects
 func (c *SnapshotRestoreController) Handle(ctx context.Context, event sdk.Event) error {
+	var (
+		snapRestore *stork_api.VolumeSnapshotRestore
+		err         error
+	)
+
 	switch o := event.Object.(type) {
 	case *stork_api.VolumeSnapshotRestore:
-		snapRestore := o
-		if snapRestore.Spec.SourceName == "" || snapRestore.Spec.SourceType == "" {
-			return fmt.Errorf("empty Snapshot name/type or namespace")
+		snapRestore = o
+		if snapRestore.Spec.SourceName == "" || snapRestore.Spec.SourceNamespace == "" {
+			return fmt.Errorf("empty Snapshot name or namespace")
 		}
 
-		if event.Deleted {
+		if snapRestore.Status.Status == stork_api.VolumeSnapshotRestoreStatusReady {
 			return nil
 		}
 
-		if snapRestore.Status.Status == stork_api.SnapshotRestoreStatusReady {
+		if event.Deleted {
+			return c.handleDelete(snapRestore)
+		}
+
+		switch snapRestore.Status.Stage {
+		case stork_api.VolumeSnapshotRestoreStageInitial:
+			err = c.handleInitial(snapRestore)
+		case stork_api.VolumeSnapshotRestoreStageRestore:
+			err = c.handleFinal(snapRestore)
+			c.Recorder.Event(snapRestore,
+				v1.EventTypeNormal,
+				string(snapRestore.Status.Status),
+				"Snapshot in-Place  Restore completed")
+		case stork_api.VolumeSnapshotRestoreStageReady:
 			return nil
+		default:
+			err = fmt.Errorf("invalid stage for volume snapshot restore: %v", snapRestore.Status.Stage)
 		}
+	}
+
+	if err != nil {
+		log.VolumeSnapshotRestoreLog(snapRestore).Errorf("Error handling event: %v err: %v", event, err.Error())
+		c.Recorder.Event(snapRestore,
+			v1.EventTypeWarning,
+			string(stork_api.VolumeSnapshotRestoreStatusError),
+			err.Error())
+	}
+
+	err = sdk.Update(snapRestore)
+	if err != nil {
+		return err
+	}
 
-		err := c.Driver.VolumeSnapshotRestore(snapRestore)
+	return nil
+}
+
+func (c *SnapshotRestoreController) handleInitial(snapRestore *stork_api.VolumeSnapshotRestore) error {
+	log.VolumeSnapshotRestoreLog(snapRestore).Infof("handled initial")
+	snapRestore.Status.Stage = stork_api.VolumeSnapshotRestoreStageRestore
+	return nil
+}
+
+func (c *SnapshotRestoreController) handleFinal(snapRestore *stork_api.VolumeSnapshotRestore) error {
+	var err error
+	// snapshot is list of snapshots
+	snapshotList := []*snap_v1.VolumeSnapshot{}
+
+	snapName := snapRestore.Spec.SourceName
+	snapNamespace := snapRestore.Spec.SourceNamespace
+	if snapRestore.Spec.GroupSnapshot {
+		logrus.Infof("GroupVolumeSnapshot In-place restore request for %v", snapName)
+		snapshotList, err = k8s.Instance().GetSnapshotsForGroupSnapshot(snapName, snapNamespace)
+		if err != nil {
+			logrus.Errorf("unable to get group snapshot details %v", err)
```

Comment:
Update all logs to use `log.VolumeSnapshotRestoreLog(snapRestore)` to help with debugging